# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Remove unneeded fields from derep object(s)
#' @param derep (`list`) [`derep`][dada2::derep-class] object, or possibly
#' nested list of such objects
#' @return None. This function is called for its side effect, which is to
#' remove the `uniques` and `quals` fields from the input object
#' @export
derep_map_only <- function(derep) {
    invisible(.Call(`_optimotu_pipeline_derep_map_only`, derep))
}

#' Convert a multiline FASTA file into a single-line FASTA file
#'
#' @param infile (`character(1)`) FASTA file, can be compressed
#' @param outfile (`character(1)`) name of output FASTA file
#' @param compress (`logical(1)`) compress output file
#' @return `character(1)` output file name
#' @export
fasta_deline <- function(infile, outfile, compress = FALSE) {
    .Call(`_optimotu_pipeline_fasta_deline`, infile, outfile, compress)
}

#' Flag the fate of reads from a FASTQ file through processing stages
#' @param raw (`character`) FASTQ file path, optionally gzipped. The raw reads.
#' @param stages (`character`) FASTQ file paths, optionally gzipped. The
#' processed reads after multiple ordered stages.
#' @return a `data.frame` with columns `read_idx` (`integer`) and `flags`
#' (`raw`) indicating which processing stages each read was found in.
#' @export
fastq_stage_flags <- function(raw, stages) {
    .Call(`_optimotu_pipeline_fastq_stage_flags`, raw, stages)
}

#' @rdname fastq_stage_flags
#' @param stage1 (`character`) FASTQ file path, optionally gzipped, the first
#' processed stage
#' @param stage2 (`character`) FASTQ file path, optionally gzipped, the second
#' processed stage
#' @export
fastq_stage_flag2 <- function(raw, stage1, stage2) {
    .Call(`_optimotu_pipeline_fastq_stage_flag2`, raw, stage1, stage2)
}

#' Map reads from a FASTQ file to processing stages
#' @param raw (`character`) FASTQ file path, optionally gzipped. The raw reads.
#' @param stages (`character`) FASTQ file paths, optionally gzipped. The
#' processed reads after multiple ordered stages.
#' @return a `data.frame` with columns `read_idx` (`integer`) and `flags`
#' (`raw`) indicating which processing stages each read was found in.
#' @export
fastq_stage_map <- function(raw, stages) {
    .Call(`_optimotu_pipeline_fastq_stage_map`, raw, stages)
}

#' Map the fate of individual reads through trimming and filtering steps
#' @param fq_raw (`character`) raw FASTQ file path, optionally gzipped
#' @param fq_trim (`character`) trimmed FASTQ file path, optionally gzipped
#' @param fq_filt (`character`) filtered FASTQ file path, optionally gzipped
#' @return `data.frame` with columns:
#' - `raw_idx` (integer) index of the sequence in fastq_raw; or if sequence names
#' in the fastq files are hex encoded integers (e.g., during subsampling to
#' indicate the original index) then `seq_id` comes from the sequence names.
#' - `trim_idx` (integer) index of the sequence in fastq_trim, or NA if absent
#' - `filt_idx` (integer) index of the sequence in fastq_filt, or NA if absent
#' @export
fastq_seq_map <- function(fq_raw, fq_trim, fq_filt) {
    .Call(`_optimotu_pipeline_fastq_seq_map`, fq_raw, fq_trim, fq_filt)
}

#' Get the names of reads in a FASTQ file
#' @param x (`character`) FASTQ file path, optionally gzipped.
#' @return (`character`) Read names.
#' @export
fastq_names <- function(x) {
    .Call(`_optimotu_pipeline_fastq_names`, x)
}

#' Get the unique quality scores in a set of FASTQ files
#' @param fastq (`character`) FASTQ file paths, optionally gzipped.
#' @param max_n (`integer`) Maximum number of reads to process.
#' @param offset (`integer`) ASCII offset for quality scores.
#' @return (`integer`) Unique quality scores.
#' @export
fastq_qual_bins <- function(fastq, max_n = 1e9L, offset = 33L) {
    .Call(`_optimotu_pipeline_fastq_qual_bins`, fastq, max_n, offset)
}

#' Interleave multiple FASTQ files
#' @param infiles (`character()`) input FASTQ files, can be compressed
#' @param outfile (`character(1)`) output FASTQ file
#' @param compress (`logical(1)`) compress output file
#' @return `character(1)` output file name
#' @export
#' @keywords internal
fastq_combine <- function(infiles, outfile, compress = FALSE) {
    .Call(`_optimotu_pipeline_fastq_combine`, infiles, outfile, compress)
}

#' Interleave multiple FASTA files
#' @param infiles (`character()`) input FASTA files, can be compressed
#' @param outfile (`character(1)`) output FASTA file
#' @param compress (`logical(1)`) compress output file
#' @return `character(1)` output file name
#' @export
#' @keywords internal
fasta_combine <- function(infiles, outfile, compress = FALSE) {
    .Call(`_optimotu_pipeline_fasta_combine`, infiles, outfile, compress)
}

#' @describeIn fastx_rename Rename reads in a FASTQ file
#' @param infile (`character`) FAST(A/Q) file path, optionally gzipped.
#' @param names (`character`) New read names.
#' @param outfile (`character`) FAST(A/Q) file path, optionally gzipped.
#' @return (`character`) Output file path.
#' @export
fastq_rename <- function(infile, names, outfile) {
    .Call(`_optimotu_pipeline_fastq_rename`, infile, names, outfile)
}

#' @describeIn fastx_rename Rename reads in a FASTA file
#' @export
fasta_rename <- function(infile, names, outfile) {
    .Call(`_optimotu_pipeline_fasta_rename`, infile, names, outfile)
}

#' Split a FASTQ file into multiple files
#' @param infile Input FASTQ file
#' @param outfiles Output FASTQ files
#' @param compress Whether to compress the output files
#' @return Output file names
#' @export
#' @keywords internal
fastq_split <- function(infile, outfiles, compress = FALSE) {
    .Call(`_optimotu_pipeline_fastq_split`, infile, outfiles, compress)
}

#' Split a FASTA file into multiple files
#' @param infile Input FASTA file
#' @param outfiles Output FASTA files
#' @param compress Whether to compress the output files
#' @return Output file names
#' @export
#' @keywords internal
fasta_split <- function(infile, outfiles, compress = FALSE) {
    .Call(`_optimotu_pipeline_fasta_split`, infile, outfiles, compress)
}

